### PR TITLE
Better handling of generic aliases

### PR DESCRIPTION
**Pull request by cdce8p**
## Steps

- [ ] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description
Followup to #921, see: https://github.com/PyCQA/astroid/pull/921#issuecomment-813937201
Improve handling of typing aliases by inserting intermediate classes in mro.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|   | :sparkles: New feature |

## Related Issue

Closes: https://github.com/PyCQA/pylint/issues/4206
Closes: https://github.com/PyCQA/pylint/issues/4284

### CODE REVIEW

File: astroid/brain/brain_typing.py [hunk 1]
```diff
@@ -161,7 +179,7 @@ def _looks_like_typing_alias(node: nodes.Call) -> bool:
         isinstance(node, nodes.Call)
         and isinstance(node.func, nodes.Name)
         and node.func.name == "_alias"
-        and isinstance(node.args[0], nodes.Attribute)
+        and isinstance(node.args[0], (nodes.Attribute, nodes.Name))
```
Comment (cdce8p) [inline astroid/brain/brain_typing.py:182]: This change is necessary, but causes many tests to fail (pylint acceptance "typing", astroid). Might be related to the brain for `builtins`.

Comment (hippo91) [reply astroid/brain/brain_typing.py:182]: @cdce8p you are right it is necessary. I found a fix for this. It will be part of my next push on #921.

Comment (cdce8p) [reply astroid/brain/brain_typing.py:182]: @hippo91 Sounds good! I'm looking forward to it.

**Comment by cdce8p** [comment]
/CC: @hippo91, @Pierre-Sassoulas

As discussed, this is the followup to #921. It addresses almost all test cases defined in https://github.com/PyCQA/pylint/pull/4239 with the exception that it can't detect the wrong usage of `re.Pattern` and `re.Match` in PY37 and 38. See example below. This would require changes to the `brain_re.py` which is out of scope for this MR. (Probably not worth it to begin with)
```py
import re
re.Pattern[str]       # [unsubscriptable-object]
re.Match[str]         # [unsubscriptable-object]
var: re.Pattern[int]  # [unsubscriptable-object]
```

File: tests/unittest_brain.py [hunk 2]
```diff
@@ -1398,12 +1420,12 @@ class Derived1(MutableSet[T]):
         """
         )
         inferred = next(node.infer())
-        check_metaclass_is_abc(inferred)
         assertEqualMro(
             inferred,
             [
                 "Derived1",
                 "MutableSet",
+                "MutableSet",
```
Comment (Pierre-Sassoulas) [inline tests/unittest_brain.py:1428]: Could this be a set with unique value ?

Comment (cdce8p) [reply tests/unittest_brain.py:1428]: It basically is. If you resolve the ClassDef with `qname`, you would get
```py
typing.MutableSet
collections.abc.MutableSet
```
`assertEqualMro` only checks the `name` of the `ClassDef`. That's why they appear the same.

Comment (Pierre-Sassoulas) [reply tests/unittest_brain.py:1428]: Ho, ok I get it now. We could do the test on `qname` so the test is more explicit then ?

Comment (cdce8p) [reply tests/unittest_brain.py]: I've updated the tests to use `qname()`

**Comment by cdce8p** [comment]
**Small update**: I've updated the brain for the `re` module. With that, **all** test cases are passing now!

**Comment by cdce8p** [comment]
If there are no objections, I would like to merge this by this weekend. After a new `astroid` release, we would finally be able to close this chapter.

**Comment by Pierre-Sassoulas** [comment]
I think @hippo91 is the one that will really understand what he's reviewing :D I did not see anything wrong with the changes myself.

File: astroid/brain/brain_builtin_inference.py [hunk 3]
```diff
@@ -153,6 +152,18 @@ def _extend_builtins(class_transforms):
...
+        builtin_name == "type"
+        and node.root().name == "re"
+        and isinstance(node.func, nodes.Name)
+        and node.func.name == "type"
+        and isinstance(node.parent, nodes.Assign)
+        and len(node.parent.targets) == 1
+        and isinstance(node.parent.targets[0], nodes.AssignName)
+        and node.parent.targets[0].name in ("Pattern", "Match")
+    ):
+        # Handle re.Pattern and re.Match in brain_re
```
Comment (hippo91) [inline astroid/brain/brain_builtin_inference.py:165]: I think a comment with an example of what is filtered could help here.

File: astroid/brain/brain_re.py [hunk 4]
```diff
@@ -34,3 +37,43 @@ def _re_transform():
...
+CLASS_GETITEM_TEMPLATE = """
+@classmethod
+def __class_getitem__(cls, item):
+    return cls
+"""
+
+
+def _looks_like_pattern_or_match(node: nodes.Call) -> bool:
+    """Check for re.Pattern or re.Match call in stdlib."""
+    return (
```
Comment (hippo91) [inline astroid/brain/brain_re.py:51]: Idem here, a simple example in the comment could help to understand

File: astroid/brain/brain_typing.py [hunk 5]
```diff
@@ -88,7 +133,13 @@ def infer_typing_attr(node, context=None):
...
         raise UseInferenceDefault from exc
 
-    if not value.qname().startswith("typing."):
+    if (
+        not value.qname().startswith("typing.")
+        or PY37
+        and value.qname() in TYPING_ALIAS
+    ):
+        # If typing subscript belongs to an alias
+        # (PY37+) handle it separately.
```
Comment (hippo91) [inline astroid/brain/brain_typing.py:142]: Well done. I do not know how i can have missed that.

File: astroid/brain/brain_typing.py [hunk 6]
```diff
@@ -174,55 +223,54 @@ def full_raiser(origin_func, attr, *args, **kwargs):
...
+    ):
+        # If typing alias is subscriptable, add `__class_getitem__` to ClassDef
+        func_to_add = astroid.extract_node(CLASS_GETITEM_TEMPLATE)
+        class_def.locals["__class_getitem__"] = [func_to_add]
+    else:
+        # If not, make sure that `__class_getitem__` access is forbidden.
+        # This is an issue in cases where the aliased class implements it,
+        # but the typing alias isn't subscriptable. E.g. `typing.ByteString` for PY39+
+        _forbid_class_getitem_access(class_def)
+    return iter([class_def])
```
Comment (hippo91) [inline astroid/brain/brain_typing.py:273]: I like very much this concise style! :+1: 
However you delete a lot of comments. It is not a pb for me if you think they where not helpful.

**Review by hippo91 — APPROVED** [review]
Well done @cdce8p . I am ok with this PR. Just left a bunch of comments. Feel free to address them or not!

File: astroid/brain/brain_typing.py (hunk 6 above)
Comment (cdce8p) [reply astroid/brain/brain_typing.py:273]: I tend to prefer less / only the necessary comments. Most information is contained in the code already. Comments also have the habit of becoming outdated pretty quickly at which point they are more confusion than actually helpful.

**Comment by cdce8p** [comment]
@hippo91 Thanks for the quick review!

**Pull request merged**